core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import cats.implicits._
+import rankn._
+
+import Identifier.Constructor
+
+sealed abstract class NormalExpression {
+  val maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(Pattern[(PackageName, Constructor), Type], NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
```

Comment:
the purpose of this would be helpful, since this is a difference from the Expr and TypeExpr graph.

Reply:
The main purpose is it allows you to be agnostic to identifiers chosen by the user that have no impact on the result of the expression. But you can accomplish this with variable renaming. It has the added value of being more agnostic to context in the normal expression.

Consider
```
foos.map(\foo -> (foo, bars.map(\bar -> bar.add(1))))
```
If you were to just do variable substitution you'd end up with something like
```
foos.map(\x0 -> (x0, bars.map(\x1 -> x1.add(1))))
```

but note that `bars.map(\bar -> bar.add(1))` could be its own expression without any unbound variables. But if you were normalize you would have
```
bars.map(\x0 -> x0.add(1))
```
Which is not what we saw in the first normalization. With de bruijn notation (I'm going to make up a bosatsu syntax that does not exist to accomplish this) we get
```
foos.map(-> (λ0, bars.map(-> λ0.add(1))))
```
and
```
bars.map(-> λ0.add(1))
```
I don't have a use for this but I could imagine someone wanting to parse the normal expressions and this would allow one more place where you can prove equality